library/ZendService/Twitter/Media.php

```diff
@@ -0,0 +1,251 @@
+use ZendService\Twitter\Response as Response;
+
+/**
+ * Twitter Media Uploader base class
+ *
+ * @author Cal Evans <cal@calevans.com>
+ */
+class Media
+{
+    const UPLOAD_BASE = 'https://upload.twitter.com';
+
+    const UPLOAD_PATH = '/1.1/media/upload.json';
+
+    /**
+     * Array of basic data to be stored by this class.
+     *
+     * @var Array
+     */
+    protected $data = [];
+
+    /**
+     * The number of bytes to send to Twitter at one time.
+     * @var Integer
+     */
+    protected $chunkSize = (1024 * 1024) * 4;
+
+    /**
+     * Constructor
+     *
+     * @param  null|string $imageFilename
+     * @param  string $consumer
+     */
+    public function __construct($imageFilename = null, $media_type = '')
+    {
+        $this->data['image_filename'] = '';
+        $this->data['media_id']       = 0;
+        $this->data['segment_index']  = 0;
+        $this->data['media_type']     = $media_type;
+
+        if (null !== $imageFilename) {
+            $this->data['image_filename'] = $imageFilename;
+        }
+    }
+
+    /**
+     * Main call into the upload workflow
+     *
+     * @param  Client $httpClient
+     * @return Twitter\Response
+     * @throws Exception If the file can't be opened.
+     */
+    public function upload(Client $httpClient)
+    {
+        if (! $this->validateFile($this->data['image_filename'])) {
+            throw new Exception\InvalidMediaException('Failed to open ' . $this->data['image_filename']);
+        }
+
+        if (empty($this->data['media_type'])) {
+            throw new Exception\InvalidMediaException('No Media Type given.');
+        }
+
+        $params = [];
+        $httpClient->setUri(self::UPLOAD_BASE . self::UPLOAD_PATH);
+
+        $params['total_bytes'] = filesize($this->data['image_filename']);
+        $params['media_type']  = $this->data['media_type'];
+
+        $response = $this->initUpload($httpClient, $params);
+        $initResponse = $response->toValue();
+
+        if (! $response->isSuccess()) {
+            throw new Exception\MediaUploadException('Failed to init the upload with Twitter.');
+        }
+
+        $this->data['media_id'] = $initResponse->media_id;
+
+        $this->appendUpload($httpClient, $params);
+
+        return $this->finalizeUpload($httpClient, $params);
+    }
+
+    /**
+     * Initalize the upload with Twitter
+     *
+     * @param Http\Client $httpClient
+     * @params array $params
+     * @return Twitter\Response
+     */
+    protected function initUpload($httpClient, $params)
+    {
+        $payload = [
+            'command'     => 'INIT',
+            'media_type'  => $params['media_type'],
+            'total_bytes' => $params['total_bytes'],
+        ];
+
+        $httpClient->resetParameters();
+        $httpClient->setHeaders(['Content-type' => 'application/x-www-form-urlencoded']);
+        $httpClient->setMethod('POST');
+        $httpClient->setParameterPost($payload);
+        $response = $httpClient->send();
```

Comment:
Refactored this method to raise an exception on error, to be consistent with the other `*Upload()` methods. As such, this line is no longer redundant; the response is tested for success and an exception raised on failure, before returning the more specific `ZendService_Twitter` response.